### PR TITLE
blog: Add IPFS experiment post

### PR DESCRIPTION
**Pull request by kallisti5**

### CODE REVIEW

**Comment by jt15s** [comment]
Any progress on this?

**Comment by kallisti5** [comment]
nope, i've been distracted :)

I'm still continuously mirroring haiku / repos / release images to ```/ipns/hpkg.haiku-os.org```

I'll try and finish this post up

**Comment by kallisti5** [comment]
Updated some of the language based on some of @nielx 's feedback.  Tried to make the language a little more common and less tech.

I think the article is a little long?  Might need trimmed down a bit.

**Comment by jt15s** [comment]
Is this ready to merge?

File: content/blog/kallisti5/2021-05-03_ipfs_experiment.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
++++
+
+I'm Alex, a member of our systems administration team and on the Haiku, Inc. board of directors.
+I have been playing with moving our repositories over to [IPFS](https://ipfs.io), and wanted to collect some feedback on the idea.
+
+## First a little history
+
+With the addition of package management in 2013, Haiku's amount of data to manage has been growing.
+
+In ~2018 I moved our Haiku package repositories (and nightly images, and release images) to S3 object storage. This helped to reduce the large amount of data we were lugging around on our core infrastructure, and offloaded it onto an externally mananged service which we could progmatically manage. All of our CI/CD could securely and pragmatically build artifacts into these S3 buckets. We found a great vendor which let us host a lot of data with unlimited egress (outbound) bandwidth for an amazing price. This worked great through 2021, however the vendor recently began walking back their "unlimited egress bandwidth" position.  In late April 2021, they shutdown our buckets resulting in a repo + nightly outage of ~24 hours while we negotiated with their support team.
```
Comment (leavengood) [inline content/blog/kallisti5/2021-05-03_ipfs_experiment.md:16]:
```suggestion
Around 2018 I moved our Haiku package repositories (and nightly images, and release images) to S3 object storage. This helped to reduce the large amount of data we were lugging around on our core infrastructure, and offloaded it onto an externally managed service which we could programmatically manage. All of our CI/CD could securely and programmatically build artifacts into these S3 buckets. We found a great vendor which let us host a lot of data with unlimited egress (outbound) bandwidth for an amazing price. This worked great through 2021, however the vendor recently began walking back their "unlimited egress bandwidth" position.  In late April 2021, they shutdown our buckets resulting in a repo + nightly outage of ~24 hours while we negotiated with their support team.
```

File: content/blog/kallisti5/2021-05-03_ipfs_experiment.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+## The downsides
+
+IPFS *is* a new technology, and there are a lot of sharp edges.
+
+* Everyone pinning needs to to manually "repin" data to pull and mirror the latest repository chunks
+* If few people are pinning the latest data, initial lookups of a CID can be a bit slow (3+ minutes)
+  * Notice, 3+ minutes is longer than the default HTTP timeout (30 seconds).
+  * Gateway timeouts can happen until the IPFS gateway "locates" the data.
+* IPFS has a steep learning curve for anyone mirroring. It takes time to find out how to do what
+* Haiku's go-lang port needs a lot more work before we can build IPFS on Haiku.
```
Comment (leavengood) [inline content/blog/kallisti5/2021-05-03_ipfs_experiment.md:73]:
```suggestion
* Haiku's Golang port needs a lot more work before we can build IPFS on Haiku.
```

File: content/blog/kallisti5/2021-05-03_ipfs_experiment.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+* Built-in signature checking ensuring security of data on IPFS
+* Deduplication is built-in, saving space
+
+Time will tell if the implementation is viable and reliable enough.
+
+In the short term, our current more traditional repositories
+are not going away as long as we can continue to host data from our S3 buckets.  I'm hopeful we can get enough
+people playing with the new system to reduce S3 bandwidth and give us some time to investigate this alternative path.
+
+A few users have mentioned adding native IPFS support to pkgman.. this would enable Haiku to obtain updates
```
Comment (leavengood) [inline content/blog/kallisti5/2021-05-03_ipfs_experiment.md:93]:
```suggestion
A few users have mentioned adding native IPFS support to pkgman...this would enable Haiku to obtain updates
```

**Review by leavengood — CHANGES_REQUESTED** [review]

**Comment by leavengood** [comment]
Let's stop holding up blog posts over minor things. I don't think this is too long. I fixed a few issues in this one with my suggestions above. If I don't hear any feedback in a few hours I will apply those and merge.

**Comment by korli** [comment]
> Let's stop holding up blog posts over minor things. I don't think this is too long. I fixed a few issues in this one with my suggestions above. If I don't hear any feedback in a few hours I will apply those and merge.

I think there is no feedfack ...  ping :)

**Comment by kallisti5** [comment]
Honestly... i'm fizzling out a bit on this.

While the idea is sound (and I used IPFS for the distribution of R1/Beta3), in actual practice at scale IPFS sucks for "big data sets".   Getting huge amounts of data pinned is unclear and unreliable from the perspective of the person trying to mirror the data.

Being able to let users "simply pin + seed our package repos to users" was really appealing.. but given the issues above it's a bit less appealing.

**Some example issues:**

* https://github.com/ipfs/go-ipfs/issues/8103
* https://github.com/ipfs/go-ipfs/issues/7586
* https://github.com/ipfs/go-ipfs/issues/3066#issuecomment-833006651
* https://github.com/ipfs/go-ipfs/issues/1702#issuecomment-670977139

I'm still mirroring our package repositories on IPFS, but not sure how hard we should push it.

tldr; The *idea* is awesome and looks like something we *badly* need.  The execution is a bit less awesome.

**Comment by jt15s** [comment]
`tldr; The idea is awesome and looks like something we badly need. The execution is a bit less awesome.`

I guess we can probably note the shortcomings of IPFS in the blog post then - it's an experiment so there are bound to be things that need improving.
